test_application/src/main.rs

```diff
@@ -0,0 +1,272 @@
+use anyhow::{self, bail};
+use clap;
+use crossbeam_channel::{select, unbounded, Receiver, Sender};
+use rand::{rngs::OsRng, thread_rng};
+use sha2::{Digest, Sha256};
+use std::{collections::HashMap, ops::Sub, thread};
+
+use tracing::{debug, info, instrument, span, trace, Level};
+use tracing_subscriber::{self, EnvFilter};
+use tss_ecdsa::{
+    AuxInfoParticipant, AuxInfoPrivate, AuxInfoPublic, Identifier, KeySharePrivate, KeySharePublic,
+    KeygenParticipant, Message, Participant, ParticipantConfig, ParticipantIdentifier,
+    PresignInput, PresignParticipant, PresignRecord, ProtocolParticipant,
+};
+
+/// Messages from main thread to router thread.
+#[derive(Debug, Copy, Clone)]
+enum SubProtocol {
+    StartKeyGen,
+    StartAuxInfo,
+    StartPresign,
+    Sign,
+}
+
+/// Outgoing message from the Participant's POV.
+#[derive(Debug)]
+enum MessageFromWorker {
+    /// Worker made progress on sub-protocol, returning some messages.
+    FinishedRound(Vec<Message>),
```

Comment:
This is great naming. It actually never occurred to me that the end of a round is exactly the point at which a participant will have messages to send out.